pkg/generators/openapi_json.go

```diff
@@ -0,0 +1,292 @@
+}
+
+func (g openAPIGeneratorJSON) generateProperty(m *types.Member, parent *types.Type) (*spec.Schema, error) {
+	var err error
+	if err := validatePatchTags(m, parent); err != nil {
+		return nil, err
+	}
+	extensions, err := g.generateExtensions(m.CommentLines)
+	if err != nil {
+		return nil, err
+	}
+	t := resolveAliasAndPtrType(m.Type)
+	jsonTags := getJsonTags(m)
+	if len(jsonTags) > 1 && jsonTags[1] == "string" {
+		schema := &spec.Schema{}
+		schema.VendorExtensible = *extensions
+		schema.SchemaProps.Type = []string{"string"}
+		return schema, nil
+	}
+	// If we can get a openAPI type and format for this type, we consider it to be simple property
+	typeString, format := openapi.GetOpenAPITypeFormat(t.String())
+	if typeString != "" {
+		// we don't call getSimpleProperty here else we'd overwrite the
+		// extensions meta above
+		schema := g.generateSimpleProperty(typeString, format)
+		schema.VendorExtensible = *extensions
+		return schema, nil
+	}
+	schema := &spec.Schema{}
+	switch t.Kind {
+	case types.Builtin:
+		return nil, fmt.Errorf("please add type %v to getOpenAPITypeFormat function", t)
+	case types.Map:
+		schema, err = g.generateMapProperty(t)
+		if err != nil {
+			return nil, err
+		}
+	case types.Slice, types.Array:
+		schema, err = g.generateSliceProperty(t)
+		if err != nil {
+			return nil, err
+		}
+	case types.Struct, types.Interface:
+		props, req, err := g.generateMembers(t, false)
+		if err != nil {
+			return nil, err
+		}
+		schema.Properties = props
+		schema.SchemaProps.Required = req
+	default:
+		return nil, fmt.Errorf("cannot generate spec for type %v", t)
+	}
+	schema.SchemaProps.Description = strings.TrimSpace(parseDescription(m.CommentLines))
+	schema.VendorExtensible = *extensions
+	return schema, nil
+}
+
+func (g openAPIGeneratorJSON) generateMapProperty(t *types.Type) (*spec.Schema, error) {
+	var err error
+	keyType := resolveAliasAndPtrType(t.Key)
+	elemType := resolveAliasAndPtrType(t.Elem)
+
+	// According to OpenAPI examples, only map from string is supported
+	if keyType.Name.Name != "string" {
+		return nil, fmt.Errorf("map with non-string keys are not supported by OpenAPI in %v", t)
+	}
+	typeString, format := openapi.GetOpenAPITypeFormat(elemType.String())
+	if typeString != "" {
+		schema := &spec.Schema{}
+		schema.Type = []string{"object"}
+		schema.AdditionalProperties = &spec.SchemaOrBool{
+			Schema: g.generateSimpleProperty(typeString, format),
+		}
+		return schema, nil
+	}
+
+	// we don't set any fields on this variable until after the switch so that
+	// we can easily call `schema, err = g.generateXYZ()` and set the fields we
+	// need at the end
+	schema := &spec.Schema{}
+	switch elemType.Kind {
+	case types.Builtin:
+		return nil, fmt.Errorf("please add type %v to getOpenAPITypeFormat function", elemType)
+	case types.Struct:
+		props, req, err := g.generateMembers(elemType, false)
+		if err != nil {
+			return nil, err
+		}
+		schema.AdditionalProperties = &spec.SchemaOrBool{
```

Comment:
also here: AdditionalProperties confuses me. Would expect `Properties`.

Reply:
@sttts As per the OpenAPI Specification for *Object Schema Model with Map/Dictionary Properties* [here](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#model-with-mapdictionary-properties) the inline members are *additional properties* merged via [this](https://github.com/OAI/OpenAPI-Specification/pull/283) PR. This is the structure of schema generated [here](https://drive.google.com/file/d/1oEn2krZASH_RZ2OaWJe8uzaMpAkRRIzp/view?usp=sharing) @munnerz would be better to answer though.

Reply:
@anubhakushwaha I think you are right. This is for maps with struct values. So the  `additionalProperties` is to be used to describe the values of unknown keys (of the map).